snuba/datasets/events_processor_base.py

```diff
@@ -0,0 +1,356 @@
+
+        # This is not ideal but it should never happen anyways
+        timestamp = _ensure_valid_date(
+            datetime.strptime(event["datetime"], settings.PAYLOAD_DATETIME_FORMAT)
+        )
+        if timestamp is None:
+            timestamp = datetime.utcnow()
+
+        output["timestamp"] = timestamp
+
+    def extract_sdk(
+        self, output: MutableMapping[str, Any], sdk: Mapping[str, Any]
+    ) -> None:
+        output["sdk_name"] = _unicodify(sdk.get("name", None))
+        output["sdk_version"] = _unicodify(sdk.get("version", None))
+
+        sdk_integrations = []
+        for i in sdk.get("integrations", None) or ():
+            i = _unicodify(i)
+            if i:
+                sdk_integrations.append(i)
+        output["sdk_integrations"] = sdk_integrations
+
+    def process_message(
+        self, message, metadata: Optional[KafkaMessageMetadata] = None
+    ) -> Optional[ProcessedMessage]:
+        """\
+        Process a raw message into a tuple of (action_type, processed_message):
+        * action_type: one of the sentinel values INSERT or REPLACE
+        * processed_message: dict representing the processed column -> value(s)
+        Returns `None` if the event is too old to be written.
+        """
+        action_type = None
+
+        if isinstance(message, dict):
+            # deprecated unwrapped event message == insert
+            action_type = ProcessorAction.INSERT
+            try:
+                processed = self.process_insert(message, metadata)
+            except EventTooOld:
+                return None
+        elif isinstance(message, (list, tuple)) and len(message) >= 2:
+            version = message[0]
+
+            if version in (0, 1, 2):
+                # version 0: (0, 'insert', data)
+                # version 1: (1, type, data, [state])
+                #   NOTE: types 'delete_groups', 'merge' and 'unmerge' are ignored
+                # version 2: (2, type, data, [state])
+                type_, event = message[1:3]
+                if type_ == "insert":
+                    action_type = ProcessorAction.INSERT
+                    try:
+                        processed = self.process_insert(event, metadata)
+                    except EventTooOld:
+                        return None
+                else:
+                    if version == 0:
+                        raise InvalidMessageType(
+                            "Invalid message type: {}".format(type_)
+                        )
+                    elif version == 1:
+                        if type_ in ("delete_groups", "merge", "unmerge"):
+                            # these didn't contain the necessary data to handle replacements
+                            return None
+                        else:
+                            raise InvalidMessageType(
+                                "Invalid message type: {}".format(type_)
+                            )
```

Comment:
We should remove these at some point.

Reply:
should we ? I think it would work for us but there may be on premise deployments where this may still be relevant

Reply:
It seems like right now we basically have an implicit guarantee that everything will be backwards compatible forever since we don't have any guarantees that it won't be.

This is obviously a pretty contrived example, but we need to figure out how to deprecate stuff like this so it doesn't slow us down when making future changes (or prevent us from being able to make them entirely.)

Reply:
(Things like the datadog configuration are other good examples, or any potential breaking changes to the command line interface.)